README.md
[![Published on NPM](https://img.shields.io/npm/v/@advanced-rest-client/paper-chip.svg)](https://www.npmjs.com/package/@advanced-rest-client/paper-chip)

[![Build Status](https://travis-ci.org/advanced-rest-client/paper-chip.svg?branch=stage)](https://travis-ci.org/advanced-rest-client/paper-chip)

[![Published on webcomponents.org](https://img.shields.io/badge/webcomponents.org-published-blue.svg)](https://www.webcomponents.org/element/advanced-rest-client/paper-chip)

# paper-chip

A compact material design element that represent and input, attribute, or action.

### Example

```html
<style>
.avatar {
  background-color: red;
  color: #fff !important;
  width: 24px;
  height: 24px;
}
</style>
<h2>Basics</h2>
<section>
  <paper-chip>Simple chip</paper-chip>

  <paper-chip>
    <span class="avatar" slot="icon">C</span>
    <span>Chip with icon</span>
  </paper-chip>

  <paper-chip removable>
    <span class="avatar" slot="icon">R</span>
    Chip that can be removed
  </paper-chip>

  <paper-chip removable disabled>
    <span class="avatar" slot="icon">D</span>
    Disabled chip
  </paper-chip>
</section>

<h2>Input chips</h2>
<section>
  <paper-chip removable>
    <iron-icon icon="communication:location-on" slot="icon"></iron-icon>
    Portland
  </paper-chip>

  <paper-chip removable>
    <iron-icon icon="maps:directions-bike" slot="icon"></iron-icon>
    Biking
  </paper-chip>
</section>
```

### API components

This components is a part of [API components ecosystem](https://elements.advancedrestclient.com/)

## Usage

### Installation
```
npm install --save @advanced-rest-client/paper-chip
```

### In an html file

```html
<html>
  <head>
    <script type="module">
      import './node_modules/@advanced-rest-client/paper-chip/paper-chip.js';
    </script>
  </head>
  <body>
    <paper-chip></paper-chip>
  </body>
</html>
```

### In a Polymer 3 element

```js
import {PolymerElement, html} from './node_modules/@polymer/polymer';
import './node_modules/@advanced-rest-client/paper-chip/paper-chip.js';

class SampleElement extends PolymerElement {
  static get template() {
    return html`
    <paper-chip></paper-chip>
    `;
  }
}
customElements.define('sample-element', SampleElement);
```

### Installation

```sh
git clone https://github.com/advanced-rest-client/paper-chip
cd api-url-editor
npm install
npm install -g polymer-cli
```

### Running the demo locally

```sh
polymer serve --npm
open http://127.0.0.1:<port>/demo/
```

### Running the tests
```sh
polymer test --npm
```

## Styling

`<paper-chip>` provides the following custom properties for styling:

Custom property | Description | Default
----------------|-------------|----------
`--paper-chip-background-color` | Chip background color | `rgba(35, 47, 52, 0.12)`
`--paper-chip-focused-background-color` | Background color when focused | `#D6D6D6`
`--paper-chip-active-background-color` | Background color when toggle is active | `#cdcdcd`
`--paper-chip-icon-color` | Color of the icon | `#666666`
`--paper-chip-label-color` | Color of the label | `#232F34`
`--paper-chip-label-focused-color` | Color of the when focused | ``
`--paper-chip-label-active-color` | Color of the when active | ``
`--paper-chip-icon-close-color` | Color of the close icon | `#dfdfdf`
`--paper-chip-icon-close-background-color` | Background color of the close icon | `#666666`

`<paper-chip>` provides the following [parts](https://www.w3.org/TR/css-shadow-parts-1/):

Part name | Description
----------------|-------------
`paper-chip-container` | Styles applied to the chip container
`paper-chip-icon` | Styles applied to the icon container
`paper-chip-label` | Styles applied to the label container
`paper-chip-remove` | Styles applied to the delete icon


CHANGELOG.md
<a name="2.0.1"></a>
## 2.0.1 (2018-11-22)


### Update

* Updated Travis configuration to connect to Sauce Labs. ([c6eec5ffb6a2106372ee002eecf251160eca26fc](https://github.com/advanced-rest-client/paper-chip/commit/c6eec5ffb6a2106372ee002eecf251160eca26fc))
* Updating styling API to enable theming ([151ba87ce84385c46a1971ef7708b5dcc9397437](https://github.com/advanced-rest-client/paper-chip/commit/151ba87ce84385c46a1971ef7708b5dcc9397437))



<a name="2.0.0"></a>
# 2.0.0 (2018-07-31)


### Update

* Updated Travis configuration to connect to Sauce Labs. ([c6eec5ffb6a2106372ee002eecf251160eca26fc](https://github.com/advanced-rest-client/paper-chip/commit/c6eec5ffb6a2106372ee002eecf251160eca26fc))



# 3.0.0 (2019-05-31)


### Breaking

* Bumping major version ([1525b30d543d372064d154ab9c59eec1d5dc3360](https://github.com/advanced-rest-client/paper-chip/commit/1525b30d543d372064d154ab9c59eec1d5dc3360))
* Upgrading to Polymer 3 ([06db40b527cf747ca83f108b654ea2a45c42617a](https://github.com/advanced-rest-client/paper-chip/commit/06db40b527cf747ca83f108b654ea2a45c42617a))

### Update

* Adding class export statement ([cb77891465fe5213494eec52332d3337498a1d95](https://github.com/advanced-rest-client/paper-chip/commit/cb77891465fe5213494eec52332d3337498a1d95))
* Disabling test coverage to test configuration ([c34d7bda41000c11d0bf0f18cd6ad24e398a9882](https://github.com/advanced-rest-client/paper-chip/commit/c34d7bda41000c11d0bf0f18cd6ad24e398a9882))
* Updated Travis configuration to connect to Sauce Labs. ([c6eec5ffb6a2106372ee002eecf251160eca26fc](https://github.com/advanced-rest-client/paper-chip/commit/c6eec5ffb6a2106372ee002eecf251160eca26fc))
* Updating styling API to enable theming ([151ba87ce84385c46a1971ef7708b5dcc9397437](https://github.com/advanced-rest-client/paper-chip/commit/151ba87ce84385c46a1971ef7708b5dcc9397437))
* Updating Travis configuration ([189e68c1bdc9371e77a43024cc7d012bc63b44fd](https://github.com/advanced-rest-client/paper-chip/commit/189e68c1bdc9371e77a43024cc7d012bc63b44fd))
* Upgrading dependencies ([904a96fbb8d1c93f91bfbd6636ba40e85974ae20](https://github.com/advanced-rest-client/paper-chip/commit/904a96fbb8d1c93f91bfbd6636ba40e85974ae20))



# 3.1.0 (2019-06-06)


### Breaking

* Bumping major version ([1525b30d543d372064d154ab9c59eec1d5dc3360](https://github.com/advanced-rest-client/paper-chip/commit/1525b30d543d372064d154ab9c59eec1d5dc3360))
* Upgrading to Polymer 3 ([06db40b527cf747ca83f108b654ea2a45c42617a](https://github.com/advanced-rest-client/paper-chip/commit/06db40b527cf747ca83f108b654ea2a45c42617a))

### New

* Adding css parts ([2f7c93e09a5fb2346950766d1529caa22f790561](https://github.com/advanced-rest-client/paper-chip/commit/2f7c93e09a5fb2346950766d1529caa22f790561))

### Update

* Adding class export statement ([cb77891465fe5213494eec52332d3337498a1d95](https://github.com/advanced-rest-client/paper-chip/commit/cb77891465fe5213494eec52332d3337498a1d95))
* Bumping minor version as API hasn't change ([3f3e41d68654bbd435be3ed23609e85b9b1435e9](https://github.com/advanced-rest-client/paper-chip/commit/3f3e41d68654bbd435be3ed23609e85b9b1435e9))
* Disabling test coverage to test configuration ([c34d7bda41000c11d0bf0f18cd6ad24e398a9882](https://github.com/advanced-rest-client/paper-chip/commit/c34d7bda41000c11d0bf0f18cd6ad24e398a9882))
* Migrating component to LitElement ([8eabf7a58bec4082dbc8a1a07e9af7ad2770b5a1](https://github.com/advanced-rest-client/paper-chip/commit/8eabf7a58bec4082dbc8a1a07e9af7ad2770b5a1))
* Updated Travis configuration to connect to Sauce Labs. ([c6eec5ffb6a2106372ee002eecf251160eca26fc](https://github.com/advanced-rest-client/paper-chip/commit/c6eec5ffb6a2106372ee002eecf251160eca26fc))
* Updating styling API to enable theming ([151ba87ce84385c46a1971ef7708b5dcc9397437](https://github.com/advanced-rest-client/paper-chip/commit/151ba87ce84385c46a1971ef7708b5dcc9397437))
* Updating Travis configuration ([189e68c1bdc9371e77a43024cc7d012bc63b44fd](https://github.com/advanced-rest-client/paper-chip/commit/189e68c1bdc9371e77a43024cc7d012bc63b44fd))
* Upgrading dependencies ([904a96fbb8d1c93f91bfbd6636ba40e85974ae20](https://github.com/advanced-rest-client/paper-chip/commit/904a96fbb8d1c93f91bfbd6636ba40e85974ae20))





demo/index.js
import { html, render } from 'lit-html';
import '@advanced-rest-client/arc-demo-helper/arc-demo-helper.js';
import '@polymer/iron-icon/iron-icon.js';
import '@polymer/iron-icons/iron-icons.js';
import '@polymer/iron-icons/communication-icons.js';
import '@polymer/iron-icons/maps-icons.js';
import '@polymer/iron-icons/device-icons.js';
import '@polymer/paper-toast/paper-toast.js';
import '../paper-chip.js';

class DemoPage {
  constructor() {
    this._toggleAmenitiesFilter = this._toggleAmenitiesFilter.bind(this);
    this._toggleTypesChoice = this._toggleTypesChoice.bind(this);
    this._handleAction = this._handleAction.bind(this);

    this.amenities = [
      {label: 'Elevator'},
      {label: 'Washer / Dryer'},
      {label: 'Fireplace'},
      {label: 'Wheelchair access'},
      {label: 'Dogs ok'},
      {label: 'Cats ok'}
    ];
  }

  get types() {
    return [
      'Extra soft',
      'Soft',
      'Medium',
      'Hard',
      'Extra hard'
    ];
  }

  _toggleAmenitiesFilter(e) {
    const index = Number(e.target.dataset.index);
    if (isNaN(index)) {
      return;
    }
    const item = Object.assign({}, this.amenities[index]);
    item.selected = !item.selected;
    this.amenities[index] = item;
    this.render();
  }

  _toggleTypesChoice(e) {
    const selectedClass = 'iron-selected';
    const selected = document.querySelector('.types .' + selectedClass);
    if (selected) {
      selected.classList.remove(selectedClass);
    }
    e.currentTarget.classList.add(selectedClass);
  }

  _handleAction(e) {
    const action = e.currentTarget.dataset.action;
    if (!action) {
      return;
    }
    const toast = document.getElementById(action + 'Action');
    toast.opened = true;
  }

  get partsTestMessage() {
    const p = document.createElement('span');
    const hasParts = p.part !== undefined;
    return hasParts ? html`<p>Your browser support CSS Shadow Parts</p>` :
      html`<p>Your browser do not support CSS Shadow Parts</p>`;
  }

  render() {
    render(html`<div class="vertical-section-container centered" role="main">
      <h2>Basics</h2>
      <section class="demo line">
        <paper-chip>
          <span>Simple chip</span>
        </paper-chip>

        <paper-chip>
          <span class="avatar" slot="icon">
            <span>C</span>
          </span>
          Chip with icon
        </paper-chip>

        <paper-chip removable="">
          <span class="avatar" slot="icon">
            <span>R</span>
          </span>
          Chip that can be removed
        </paper-chip>

        <paper-chip removable="" disabled="">
          <span class="avatar" slot="icon">
            <span>D</span>
          </span>
          Disabled chip
        </paper-chip>
      </section>

      <h2>Input chips</h2>
      <section class="demo line">
        <paper-chip removable="">
          <iron-icon icon="communication:location-on" slot="icon"></iron-icon>
          Portland
        </paper-chip>

        <paper-chip removable="">
          <iron-icon icon="maps:directions-bike" slot="icon"></iron-icon>
          Biking
        </paper-chip>
      </section>

      <h2>Choice chips</h2>
      <section class="demo">
        <h3>Select type</h3>
        <div class="wrap-horizontal types">
        ${this.types.map((item) => html`<paper-chip @click="${this._toggleTypesChoice}">
          ${item}
        </paper-chip>`)}
        </div>
      </section>

      <h2>Filter chips</h2>
      <section class="demo">
        <h3>Choose amenities</h3>
        <div class="wrap-horizontal">
          ${this.amenities.map((item, index) => html`<paper-chip @click="${this._toggleAmenitiesFilter}"
            toggles data-index="${index}">
            ${item.selected ? html`<iron-icon icon="check" slot="icon"></iron-icon>` : ''}
            ${item.label}
          </paper-chip>`)}
        </div>
      </section>

      <h2>Action chips</h2>
      <section class="demo">
        <div class="wrap-horizontal actions">
          <paper-chip @click="${this._handleAction}" data-action="brightness">
            <iron-icon icon="device:brightness-low" slot="icon"></iron-icon>
            <span>Turn on lights</span>
          </paper-chip>

          <paper-chip @click="${this._handleAction}" data-action="alarm">
            <iron-icon icon="alarm" slot="icon"></iron-icon>
            <span>Set alarm</span>
          </paper-chip>

          <paper-chip @click="${this._handleAction}" data-action="clear">
            <iron-icon icon="communication:clear-all" slot="icon"></iron-icon>
            <span>Clear all</span>
          </paper-chip>
        </div>
      </section>

      <h2>Themed chips</h2>
      <section class="demo">
        <div class="wrap-horizontal actions">
          <div class="themed">
            <paper-chip>
              <iron-icon icon="device:brightness-low" slot="icon"></iron-icon>
              Styled text child
            </paper-chip>

            <paper-chip>
              <iron-icon icon="device:brightness-low" slot="icon"></iron-icon>
              <span>Styled element child</span>
            </paper-chip>
          </div>
        </div>
      </section>

      <h2>Themed chips with CSS parts</h2>
      ${this.partsTestMessage}
      <section class="demo">
        <div class="wrap-horizontal actions">
          <div class="themed-parts">
            <paper-chip removable>
              <iron-icon icon="device:brightness-low" slot="icon"></iron-icon>
              Styled with CSS parts
            </paper-chip>

            <paper-chip>
              <iron-icon icon="device:brightness-low" slot="icon"></iron-icon>
              <span>Also styled with parts</span>
            </paper-chip>
          </div>
        </div>
      </section>
    </div>

    <paper-toast id="brightnessAction" text="Turing lights on"></paper-toast>
    <paper-toast id="alarmAction" text="Setting the alarm"></paper-toast>
    <paper-toast id="clearAction" text="Clearing all actions"></paper-toast>`, document.querySelector('#demo'));
  }
}

window.addEventListener('chip-removed', (e) => {
  e.target.parentNode.removeChild(e.target);
});

const instance = new DemoPage();
instance.render();
window._demo = instance;


PaperChip.d.ts
/**
 * DO NOT EDIT
 *
 * This file was automatically generated by
 *   https://github.com/Polymer/tools/tree/master/packages/gen-typescript-declarations
 *
 * To modify these typings, edit the source file(s):
 *   PaperChip.js
 */


// tslint:disable:variable-name Describing an API that's defined elsewhere.
// tslint:disable:no-any describes the API as best we are able today

import {LitElement, html, css} from 'lit-element';

export {PaperChip};

declare namespace UiElements {

  /**
   * `paper-chip`
   *
   * A compact material design element that represent and input, attribute, or action.
   *
   * A chip contains a label and optionally an icon and remove icon.
   *
   * Remove icon is predefined. However icon can be any HTML element with
   * `slot="icon"` attribute. Per material design guidelines the icon is rounded.
   *
   * ## Example
   *
   * ```html
   * <paper-chip removable>
   *  <iron-icon icon="maps:directions-bike" slot="icon"></iron-icon>
   *  Biking
   * </paper-chip>
   * ```
   *
   * The "Bliking" is the label rendered next to the icon. The chip also renders
   * built-in remove icon. Clicking on the icon dispatches `chip-removed`
   * custom event only. It does not remove the chip from the document as the
   * application logic might use different ways of removing elements from dom
   * than web platform APIs.
   *
   * ## Styling
   *
   * `<paper-chip>` provides the following custom properties for styling:
   *
   * Custom property | Description | Default
   * ----------------|-------------|----------
   * `--paper-chip-background-color` | Chip background color | `rgba(35, 47, 52, 0.12)`
   * `--paper-chip-focused-background-color` | Background color when focused | `#D6D6D6`
   * `--paper-chip-active-background-color` | Background color when toggle is active | `#cdcdcd`
   * `--paper-chip-icon-color` | Color of the icon | `#666666`
   * `--paper-chip-label-color` | Color of the label | `#232F34`
   * `--paper-chip-label-focused-color` | Color of the when focused | ``
   * `--paper-chip-label-active-color` | Color of the when active | ``
   * `--paper-chip-icon-close-color` | Color of the close icon | `#dfdfdf`
   * `--paper-chip-icon-close-background-color` | Background color of the close icon | `#666666`
   *
   * `<paper-chip>` provides the following [parts](https://www.w3.org/TR/css-shadow-parts-1/):
   *
   * Part name | Description
   * ----------------|-------------
   * `paper-chip-container` | Styles applied to the chip container
   * `paper-chip-icon` | Styles applied to the icon container
   * `paper-chip-label` | Styles applied to the label container
   * `paper-chip-remove` | Styles applied to the delete icon
   */
  class PaperChip extends LitElement {

    /**
     * If true, the user cannot interact with this element.
     */
    disabled: Boolean|null;
    readonly active: Boolean|null;
    _active: any;
    readonly focused: Boolean|null;
    _focused: any;
    readonly _iconSlot: HTMLElement|null;

    /**
     * If set the chip can be removed.
     * The element does not remove itself from the DOM. It rather dispatch
     * `chip-removed` custom event to inform parent element about the action.
     */
    removable: boolean|null|undefined;

    /**
     * A name of the icon to render when `removable` property is set.
     * By default it referes to Polymer's default icons library, to the
     * `clear` icon. You must include this library into your document.
     * You can also use whatever other icons library.
     */
    removeIcon: string|null|undefined;
    _hasIconNode: boolean|null|undefined;

    /**
     * If true, the button toggles the active state with each click or press
     * of the spacebar or enter.
     */
    toggles: boolean|null|undefined;
    constructor();
    connectedCallback(): void;
    disconnectedCallback(): void;
    firstUpdated(changedProperties: any): void;
    render(): any;

    /**
     * Adds the `slotchange` event listener to the icon slot.
     */
    _addSlotEvent(): void;

    /**
     * Handler for remove icon click event.
     * Cancels the event and calls `remove()`
     */
    _removeHandler(e: ClickEvent|null): void;

    /**
     * Dispatches `chip-remove` custom event to inform application
     * that the user requested to remove the item.
     *
     * Note, this does not check if `removable` is set.
     */
    remove(): void;

    /**
     * According to material design spec, when there's no icon the
     * left hand side padding should be 12dp. Slotted styling API does now
     * allow to detect when there's no contect so it has to be done using
     * node observer.
     */
    _detectHasIcon(): void;

    /**
     * Computes class name for the container.
     *
     * @param hasIconNode True if the element has an icon in the light DOM.
     * @param removable True if the element can be removed.
     * @returns Class name.
     */
    _computeContainerClass(hasIconNode: Boolean|null, removable: Boolean|null): String|null;

    /**
     * Handler for key down when element is focused.
     * Removes the item when delete key is pressed.
     */
    _keyDownHandler(e: KeyboardEvent|null): void;

    /**
     * Sets state of the `focused` property depending on the event handled by this
     * listener.
     *
     * @param e Either focus or blur events
     */
    _focusBlurHandler(e: Event|null): void;

    /**
     * Calles when the value of `disabled` property change. Sets `aria-disabled`
     * and `tabIndex` attributes.
     *
     * @param disabled Current value of `disabled` property.
     */
    _disabledChanged(disabled: Boolean|null): void;

    /**
     * Handles click event (as well as Space and Enter key down) as sets the
     * `active` property.
     */
    _clickHandler(): void;

    /**
     * Sets `_active` property depending on the input and current state of `_active`.
     *
     * @param active The value to set.
     */
    _userActivate(active: Boolean|null): void;

    /**
     * Calls `clic()` function on this element so event listeners can handle
     * the action.
     */
    _asyncClick(): void;
  }
}


PaperChip.js
import { LitElement, html, css } from 'lit-element';
import '@polymer/iron-icon/iron-icon.js';
/**
 * `paper-chip`
 *
 * A compact material design element that represent and input, attribute, or action.
 *
 * A chip contains a label and optionally an icon and remove icon.
 *
 * Remove icon is predefined. However icon can be any HTML element with
 * `slot="icon"` attribute. Per material design guidelines the icon is rounded.
 *
 * ## Example
 *
 * ```html
 * <paper-chip removable>
 *  <iron-icon icon="maps:directions-bike" slot="icon"></iron-icon>
 *  Biking
 * </paper-chip>
 * ```
 *
 * The "Bliking" is the label rendered next to the icon. The chip also renders
 * built-in remove icon. Clicking on the icon dispatches `chip-removed`
 * custom event only. It does not remove the chip from the document as the
 * application logic might use different ways of removing elements from dom
 * than web platform APIs.
 *
 * ## Styling
 *
 * `<paper-chip>` provides the following custom properties for styling:
 *
 * Custom property | Description | Default
 * ----------------|-------------|----------
 * `--paper-chip-background-color` | Chip background color | `rgba(35, 47, 52, 0.12)`
 * `--paper-chip-focused-background-color` | Background color when focused | `#D6D6D6`
 * `--paper-chip-active-background-color` | Background color when toggle is active | `#cdcdcd`
 * `--paper-chip-icon-color` | Color of the icon | `#666666`
 * `--paper-chip-label-color` | Color of the label | `#232F34`
 * `--paper-chip-label-focused-color` | Color of the when focused | ``
 * `--paper-chip-label-active-color` | Color of the when active | ``
 * `--paper-chip-icon-close-color` | Color of the close icon | `#dfdfdf`
 * `--paper-chip-icon-close-background-color` | Background color of the close icon | `#666666`
 *
 * `<paper-chip>` provides the following [parts](https://www.w3.org/TR/css-shadow-parts-1/):
 *
 * Part name | Description
 * ----------------|-------------
 * `paper-chip-container` | Styles applied to the chip container
 * `paper-chip-icon` | Styles applied to the icon container
 * `paper-chip-label` | Styles applied to the label container
 * `paper-chip-remove` | Styles applied to the delete icon
 *
 * @customElement
 * @polymer
 * @demo demo/index.html
 * @memberof UiElements
 */
export class PaperChip extends LitElement {
  static get styles() {
    return css`:host {
      display: inline-block;
      outline: none;
      cursor: default;
      margin: 4px;
      box-sizing: border-box;
    }

    .container {
      border-radius: 16px;
      background-color: var(--paper-chip-background-color, rgba(35, 47, 52, 0.12));
      height: inherit;
      min-height: 32px;
      -moz-user-select: none;
      -webkit-user-select: none;
      -ms-user-select: none;
      user-select: none;
      transition: box-shadow 0.28s cubic-bezier(0.4, 0, 0.2, 1);

      display: flex;
      -ms-flex-direction: row;
      flex-direction: row;
      -ms-flex-align: center;
      align-items: center;
    }

    :host([focused]) .container {
      box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.14),
                  0 1px 5px 0 rgba(0, 0, 0, 0.12),
                  0 3px 1px -2px rgba(0, 0, 0, 0.2);
      background-color: var(--paper-chip-focused-background-color, #D6D6D6);
    }

    :host([active]) .container {
      background-color: var(--paper-chip-active-background-color, #cdcdcd);
    }

    :host([toggles]) {
      cursor: pointer;
    }

    :host([disabled]) {
      opacity: 0.54;
      pointer-events: none;
    }

    .icon ::slotted([slot=icon]) {
      border-radius: 50%;
      margin: 4px 0 4px 6px;
      color: var(--paper-chip-icon-color, #666666);
    }

    .label {
      display: inline-block;
      padding: 0px 8px;
      margin-left: 12px;
      margin-right: 12px;
      font-size: var(--arc-font-body2-font-size);
      font-weight: var(--arc-font-body2-font-weight);
      line-height: var(--arc-font-body2-line-height);
      color: var(--paper-chip-label-color, #232F34);
    }

    .label ::slotted([slot]) {
      font-size: var(--arc-font-body2-font-size);
      font-weight: var(--arc-font-body2-font-weight);
      line-height: var(--arc-font-body2-line-height);
      color: var(--paper-chip-label-color, #232F34);
    }

    :host([focused]) ::slotted([slot]),
    :host([focused]) .label {
      color: var(--paper-chip-label-focused-color);
    }

    :host([active]) ::slotted([slot]),
    :host([active]) .label {
      color: var(--paper-chip-label-active-color);
    }

    .with-icon .label {
      margin-left: 0;
    }

    .with-remove .label {
      margin-right: 0;
    }

    .close {
      width: 16px;
      height: 16px;
      background-color: var(--paper-chip-icon-close-background-color, #666666);
      color: var(--paper-chip-icon-close-color, #dfdfdf);
      border-radius: 50%;
      margin-right: 6px;
      cursor: pointer;
    }`;
  }

  static get properties() {
    return {
      /**
       * If set the chip can be removed.
       * The element does not remove itself from the DOM. It rather dispatch
       * `chip-removed` custom event to inform parent element about the action.
       */
      removable: { type: Boolean },
      /**
       * A name of the icon to render when `removable` property is set.
       * By default it referes to Polymer's default icons library, to the
       * `clear` icon. You must include this library into your document.
       * You can also use whatever other icons library.
       */
      removeIcon: { type: String },

      _hasIconNode: { type: Boolean },
      /**
       * If true, the user cannot interact with this element.
       */
      disabled: { type: Boolean, reflect: true },
      /**
       * If true, the button toggles the active state with each click or press
       * of the spacebar or enter.
       */
      toggles: { type: Boolean, reflect: true }
    };
  }

  set disabled(value) {
    this.__disabled = value;
    this._disabledChanged(value);
  }
  /**
   * @return {Boolean} True if the button is in disabled state.
   */
  get disabled() {
    return this.__disabled;
  }
  /**
   * @return {Boolean} True if the button is currently in active state. Only
   * available if the button `toggles`.
   */
  get active() {
    return this.__active;
  }

  get _active() {
    return this.__active;
  }

  set _active(value) {
    this.__active = value;
    this.__activeChanged(value);
  }
  /**
   * @return {Boolean} True if the button is currently in focused state.
   */
  get focused() {
    return this.__focused || false;
  }

  get _focused() {
    return this.__focused;
  }

  set _focused(value) {
    this.__focused = value;
    if (value && !this.hasAttribute('focused')) {
      this.setAttribute('focused', '');
    } else if (!value && this.hasAttribute('focused')) {
      this.removeAttribute('focused');
    }
  }

  render() {
    const containerClass = this._computeContainerClass(this._hasIconNode, this.removable);
    return html`<div part="paper-chip-container" class="container ${containerClass}">
      <span part="paper-chip-icon" class="icon"><slot name="icon"></slot></span>
      <span part="paper-chip-label" class="label"><slot></slot></span>
      ${this.removable ? html`<iron-icon part="paper-chip-remove"
        class="close" icon="${this.removeIcon}" @click="${this._removeHandler}"></iron-icon>` : ''}
    </div>`;
  }
  /**
   * @return {HTMLElement} Reference to the icon slot element.
   */
  get _iconSlot() {
    return this.shadowRoot.querySelector('slot[name="icon"]');
  }

  constructor() {
    super();
    this.removeIcon = 'clear';
    this._keyDownHandler = this._keyDownHandler.bind(this);
    this._focusBlurHandler = this._focusBlurHandler.bind(this);
    this._clickHandler = this._clickHandler.bind(this);
  }

  connectedCallback() {
    super.connectedCallback();
    if (!this.hasAttribute('tabindex')) {
      this.setAttribute('tabindex', '0');
    }
    if (!this.hasAttribute('role')) {
      this.setAttribute('role', 'button');
    }
    /* istanbul ignore else */
    if (this.__active === undefined) {
      this._active = false;
    }
    this._addSlotEvent();
    this.addEventListener('keydown', this._keyDownHandler, true);
    this.addEventListener('focus', this._focusBlurHandler, true);
    this.addEventListener('blur', this._focusBlurHandler, true);
    this.addEventListener('click', this._clickHandler, true);
  }

  disconnectedCallback() {
    super.disconnectedCallback();
    this.removeEventListener('keydown', this._keyDownHandler);
    this.removeEventListener('focus', this._focusBlurHandler);
    this.removeEventListener('blur', this._focusBlurHandler);
    this.removeEventListener('click', this._clickHandler);
  }

  firstUpdated(changedProperties) {
    super.firstUpdated(changedProperties);
    this.__firstUpdated = true;
    this._addSlotEvent();
  }

  /**
   * Adds the `slotchange` event listener to the icon slot.
   */
  _addSlotEvent() {
    if (!this.__firstUpdated) {
      return;
    }
    this._iconSlot.addEventListener('slotchange', () => this._detectHasIcon());
  }

  /**
   * Handler for remove icon click event.
   * Cancels the event and calls `remove()`
   *
   * @param {ClickEvent} e
   */
  _removeHandler(e) {
    e.preventDefault();
    e.stopPropagation();
    this.remove();
  }

  /**
   * Dispatches `chip-remove` custom event to inform application
   * that the user requested to remove the item.
   *
   * Note, this does not check if `removable` is set.
   */
  remove() {
    this._active = false;
    this.dispatchEvent(new CustomEvent('chip-removed', {
      composed: true,
      bubbles: true
    }));
  }
  /**
   * According to material design spec, when there's no icon the
   * left hand side padding should be 12dp. Slotted styling API does now
   * allow to detect when there's no contect so it has to be done using
   * node observer.
   */
  _detectHasIcon() {
    const nodes = this._iconSlot.assignedNodes()
      .filter((node) => node.nodeType === Node.ELEMENT_NODE);
    this._hasIconNode = !!nodes.length;
  }
  /**
   * Computes class name for the container.
   * @param {Boolean} hasIconNode True if the element has an icon in the light DOM.
   * @param {Boolean} removable True if the element can be removed.
   * @return {String} Class name.
   */
  _computeContainerClass(hasIconNode, removable) {
    let klass = '';
    if (hasIconNode) {
      klass += 'with-icon';
    }
    if (removable) {
      if (klass) {
        klass += ' ';
      }
      klass += 'with-remove';
    }
    return klass;
  }
  /**
   * Handler for key down when element is focused.
   * Removes the item when delete key is pressed.
   * @param {KeyboardEvent} e
   */
  _keyDownHandler(e) {
    if (this.removable && (e.key === 'Backspace' || e.key === 'Delete')) {
      this.remove();
    } else if (e.key === ' ' || e.key === 'Enter') {
      this._clickHandler();
      this._asyncClick();
    }
  }
  /**
   * Sets state of the `focused` property depending on the event handled by this
   * listener.
   * @param {Event} e Either focus or blur events
   */
  _focusBlurHandler(e) {
    this._focused = e.type === 'focus';
  }
  /**
   * Calles when the value of `disabled` property change. Sets `aria-disabled`
   * and `tabIndex` attributes.
   * @param {Boolean} disabled Current value of `disabled` property.
   */
  _disabledChanged(disabled) {
    this.setAttribute('aria-disabled', disabled ? 'true' : 'false');
    this.style.pointerEvents = disabled ? 'none' : '';
    if (disabled) {
      // Read the `tabindex` attribute instead of the `tabIndex` property.
      // The property returns `-1` if there is no `tabindex` attribute.
      // This distinction is important when restoring the value because
      // leaving `-1` hides shadow root children from the tab order.
      this._oldTabIndex = this.getAttribute('tabindex');
      this._focused = false;
      this.setAttribute('tabindex', -1);
      this.blur();
    } else if (this._oldTabIndex !== undefined) {
      if (this._oldTabIndex === null) {
        this.removeAttribute('tabindex');
      } else {
        this.setAttribute('tabindex', this._oldTabIndex);
      }
    }
  }
  /**
   * Handles click event (as well as Space and Enter key down) as sets the
   * `active` property.
   */
  _clickHandler() {
    if (this.toggles) {
      this._userActivate(!this._active);
    } else if (this._active) {
      this._active = false;
    }
  }
  /**
   * Sets `_active` property depending on the input and current state of `_active`.
   * @param {Boolean} active The value to set.
   */
  _userActivate(active) {
    if (this._active !== active) {
      this._active = active;
    }
  }
  /**
   * Calls `clic()` function on this element so event listeners can handle
   * the action.
   */
  _asyncClick() {
    setTimeout(() => {
      this.click();
    }, 1);
  }
  /**
   * Called when the `active` value change.
   * It sets `active` attribute and, if the button toggles, `aria-pressed` attribute.
   * @param {Boolean} active Current state of `active`./
   */
  __activeChanged(active) {
    if (active) {
      if (!this.hasAttribute('active')) {
        this.setAttribute('active', '');
      }
    } else {
      if (this.hasAttribute('active')) {
        this.removeAttribute('active');
      }
    }
    if (this.toggles) {
      this.setAttribute('aria-pressed', active ? 'true' : 'false');
    } else {
      if (this.hasAttribute('aria-pressed')) {
        this.removeAttribute('aria-pressed');
      }
    }
  }
  /**
   * Dispatched when the user clicked on "remove" button or when `remove()`
   * function is called.
   * The event bubbles through shadow DOM.
   *
   * @event chip-removed
   */
}
